airflow/timetables/events.py

```diff
@@ -79,22 +83,34 @@ def next_dagrun_info(
         last_automated_data_interval: DataInterval | None,
         restriction: TimeRestriction,
     ) -> DagRunInfo | None:
-        if last_automated_data_interval is None:
-            next_event = self.event_dates[0]
-        else:
-            future_dates = itertools.dropwhile(
-                lambda when: when <= last_automated_data_interval.end,  # type: ignore
-                self.event_dates,
-            )
-            next_event = next(future_dates, None)  # type: ignore
-            if next_event is None:
-                return None
+        earliest = restriction.earliest
+        if not restriction.catchup:
+            current_time = timezone.utcnow()
+            if earliest is None or current_time > earliest:
+                earliest = pendulum.instance(current_time)
+
+        dates = iter(self.event_dates)
+        next_event = next(dates, None)
+        while next_event:
+            is_allowed = True
+            if earliest and next_event < earliest:
+                is_allowed = False
+            if last_automated_data_interval and next_event <= last_automated_data_interval.end:
+                is_allowed = False
+            if is_allowed:
+                break
+            next_event = next(dates, None)
+        if next_event is None:
+            return None
```

Comment:
```suggestion
        if not self.event_dates:
            return None
        for next_event in self.event_dates:
            if earliest and next_event < earliest:
                continue
            elif last_automated_data_interval and next_event <= last_automated_data_interval.end:
                continue
            break
```

Not exactly sure but this seems to be logically equivalent?